### PR TITLE
Moving CollectionExtensions to System.Linq.Expressions

### PR DESCRIPTION
**Pull request by bartdesmet**
With the elimination of `System.Dynamic.Runtime` as a separate assembly, some utilities can be moved out of the `Common` folder. Feeling the CI waters here by moving `CollectionExtensions`. If successful, more PRs starring such moves will follow.

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs [hunk 1]
```diff
@@ -3,17 +3,129 @@
...
     {
+        public static T[] AddFirst<T>(this IList<T> list, T item)
+        {
+            T[] res = new T[list.Count + 1];
+            res[0] = item;
+            list.CopyTo(res, 1);
+            return res;
+        }
+
         public static T[] AddLast<T>(this IList<T> list, T item)
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]: Noticed that this only operates on arrays, so we can change the type from `IList<T>` to `T[]` at some point. Will do that in a separate PR.

**Comment by JonHanna** [comment]
Ha! I just turned on my computer to do precisely this 😄

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs [hunk 2]
```diff
@@ -3,17 +3,129 @@
...
+        }
+
+        /// <summary>
+        /// Wraps the provided enumerable into a ReadOnlyCollection{T}
+        ///
+        /// Copies all of the data into a new array, so the data can't be
+        /// changed after creation. The exception is if the enumerable is
+        /// already a ReadOnlyCollection{T}, in which case we just return it.
+        /// </summary>
+        [Pure]
```
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]: The is about the only type with this attribute used on it. Does it give any benefit with such limited use?

Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]: Yeah, I think we can remove these remnants of contracts. @VSadov, @stephentoub what do you think? Is there still any value in keeping things like `[Pure]` around?

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs [hunk 3]
```diff
@@ -3,17 +3,129 @@
...
+
+        public static T[] RemoveLast<T>(this T[] array)
+        {
+            T[] result = new T[array.Length - 1];
+            Array.Copy(array, 0, result, 0, result.Length);
+            return result;
+        }
+
+        // Name needs to be different so it doesn't conflict with Enumerable.Select
+        public static U[] Map<T, U>(this T[] array, Func<T, U> select)
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]: Does the delegate type matter for usage of Map?  I ask because if it doesn't, since this method is internal, it could be removed entirely and replaced with usage of Array.ConvertAll, which does the same thing with the same signature, other than the type of the delegate.

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]: I'm fine with removing them.

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs (hunk 3 above)
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/CollectionExtensions.cs]: Good point. I'll look into it.
